susemanager-utils/susemanager-sls/susemanager-sls.spec

```diff
@@ -141,28 +141,52 @@ base:
 EOF
     fi
 fi
-# Restrict Java RMI to localhost (bsc#1184617)
-restrict_to_localhost()
+# Run JMX exporter as Java Agent (bsc#1184617)
+set_up_java_agent()
 {
-  JMXREMOTE_HOST='-Dcom.sun.management.jmxremote.host='
-  JMXREMOTE_PORT='-Dcom.sun.management.jmxremote.port='
-  RMI_SERVER_HOSTNAME='-Djava.rmi.server.hostname='
-  systemctl is-enabled ${1} > /dev/null 2>&1
+  JMXREMOTE_OPT='-Dcom\.sun\.management\.jmxremote\.'
+  RMI_SERVER_HOSTNAME='-Djava\.rmi\.server\.hostname='
+  JMX_EXPORTER_JAVA_AGENT='jmx_prometheus_javaagent.jar'
+  service_name='prometheus-jmx_exporter@'${1}'.service'
+  systemctl is-enabled $service_name > /dev/null 2>&1
   jmx_exporter_enabled=$?
-  grep -q -- $JMXREMOTE_HOST ${2}
-  jmxremote_host_configured=$?
-  if [ $jmx_exporter_enabled -eq 0 ] && [ $jmxremote_host_configured -eq 1 ]; then
-    sed -ri "s/JAVA_OPTS=\"(.*)${JMXREMOTE_PORT}(.*)\"/JAVA_OPTS=\"\1${JMXREMOTE_HOST}localhost\ ${JMXREMOTE_PORT}\2\"/" ${2}
-    sed -ri "s/JAVA_OPTS=\"(.*)${RMI_SERVER_HOSTNAME}\S*(.*)\"/JAVA_OPTS=\"\1${RMI_SERVER_HOSTNAME}localhost\2\"/" ${2}
+  grep -q -- $JMXREMOTE_OPT ${2}
+  jmxremote_opt_configured=$?
+  grep -q -- $JMX_EXPORTER_JAVA_AGENT ${2}
+  jmx_exporter_java_agent_configured=$?
+  if [ $jmx_exporter_enabled -eq 0 ]; then
+    systemctl disable $service_name
```

Comment:
IIUC this disables the old JMX-based service if it exists.

I wonder why this was implemented as a post script instead of via Salt states, what pros and cons do you see?

Reply:
The post script and Salt states fulfill two different tasks.
Salt states allow interactively enabling and disabling monitoring setup and check for the current status.
The post script is meant to be run only once and migrate the setup to the new configuration in case monitoring has already been enabled in the old configuration.